### PR TITLE
[receiver/k8sobjects] Fix memory leak caused by pull mode's interval ticker

### PR DESCRIPTION
**Pull request by crobert-1**
**Description:** <Describe what has changed.>

The k8sobjects receiver was starting a ticker when in pull mode, that would tick every set interval. This ticker needs to be stopped during shutdown, so I added functionality to cancel the context during shutdown, properly stopping the ticker.

The original intention here was to add `goleak`, but due to a bug in `internal/k8stest` it's failing on the end to end test. This change still fixes a bug, so I believe it's worth merging even without `goleak` at this point. Note that this change does reduce the number and frequency of goleak failures.

**Testing:** <Describe what testing was performed and which tests were added.>
All existing tests are passing.

### CODE REVIEW

File: .chloggen/goleak_k8sobjectsreceiver.yaml [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. filelogreceiver)
+component: k8sobjectsreceiver
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Fix memory leak caused by the pull mode's interval ticker
+
+# Mandatory: One or more tracking issues related to the change. You can use the PR number here if no issue exists.
+issues: []
```
Comment (crobert-1) [inline .chloggen/goleak_k8sobjectsreceiver.yaml:13]:
```suggestion
issues: [31919]
```

**Comment by crobert-1** [comment]
I'll investigate failing goleak test and mark as `Ready to Review` when I've resolved it.

**Comment by TylerHelmuth** [comment]
I wonder how related this is to https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/31644

**Comment by crobert-1** [comment]
> I wonder how related this is to #31644

I suspect this is actually frequency of https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/31047, since the only test that's failing is using k8stest, and the leaked goroutines have the same stack.

**Comment by crobert-1** [comment]
I've updated the issue description. `goleak` is still failing on the end to end test due to a bug in `internal/k8stest`. I believe this is still worth merging to fix the memory leak detected, even if `goleak` isn't included.

**Review by TylerHelmuth — APPROVED** [review]

**Review by TylerHelmuth — APPROVED** [review]

**Pull request merged**